cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -1382,6 +1385,76 @@ func maybeMountSave(disk disks.Disk, rootdir string, encrypted bool, mountOpts *
 	return true, nil
 }
 
+func generateMountsModeClassic(mst *initramfsMountsState, label string) error {
+	// Mount ESP
+	// Must be where the kernel came from, without fallbacks
+	espuuid, err := bootFindPartitionUUIDForBootedKernelDisk()
+	if err != nil {
+		return err
+	}
+	espSrc := filepath.Join("/dev/disk/by-partuuid", espuuid)
+	espOpts := &systemdMountOptions{
+		// vfat needs fsck to mount ESP that was unsafely
+		// unmounted
+		NeedsFsck: true,
+		// do not persist mount into running system, otherwise
+		// regular fstab cannot mount ESP at /boot/efi as
+		// expected
+		Ephemeral: true,
+	}
+	if err := doSystemdMount(espSrc, boot.InitramfsUbuntuSeedDir, espOpts); err != nil {
+		return err
+	}
+
+	// get the disk that we mounted the ESP from as a reference
+	// point for future mounts
+	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+	if err != nil {
+		return err
+	}
+
+	// TODO measure model, if we will have one
+
+	// Mount rootfs
+	if err := secbootTpmPrepare(boot.InitramfsUbuntuSeedDir); err != nil {
+		return err
+	}
+	runModeKey := filepath.Join(boot.InitramfsSeedEncryptionKeyDir, label+".sealed-key")
+	opts := &secboot.UnlockVolumeUsingSealedKeyOptions{
+		AllowRecoveryKey: true,
```

Comment:
is it really a supported use case to allow entering the recovery key for a cloud image? How would that work? Is it a common thing for clouds to have interactive consoles? I ask because I don't think GCE does or at least we have had that problem with GCE before

Reply:
Most clouds do, and it is explicit requirement.

Usually one wouldn't do that via serial console, but by detaching disk and moving it to a different VM with a different TPM, meaning one will ssh-in normally but will be unable to unlock the disk as the TPM is different.